python/diff_frames.py

```diff
@@ -25,36 +108,116 @@ def duration(sndfile):
     return sndfile.frames() / sndfile.samplerate()
 
 
-def fmt_seconds(t):
-    hours, remainder = divmod(t, 3600)
-    minutes, seconds = divmod(remainder, 60)
-    if not hours:
-        if not minutes:
-            return '{:.2f}'.format(seconds)
-        return '{:.0f}:{:.1f}'.format(minutes, seconds)
-    return '{:.0f}:{:.0f}:{.0f}'.format(hours, minutes, seconds)
+class _DrawState:
+    '''Throw-away objects for ease of re-initialising draw state on every draw()
+    call.
+    '''
+    __slots__ = (
+        'app', '_diff_width', '_chars_per_frame', 'start_times', 'end_times')
+
+    def __init__(self, app):
+        self.app = app
+        self.start_times = None
+        self.end_times = None
+
+    @property
+    @cache
+    def diff_width(self):
+        if self.start_times:
+            start_time_width = max(map(len, self.start_times)) + 1
+        else:
+            start_time_width = 0
+        end_time_width = max(map(len, self.end_times)) + 1
+        return self.app._terminal.width - start_time_width - end_time_width
+
+    @property
+    @cache
+    def chars_per_frame(self):
+        return self.app._zoom * self.diff_width / self.app._len
+
+    def to_chars(self, frames):
+        '''Transforms a point in the space of the "virtual" diff space, given in
+        frames, into a x-coordintate on the terminal, taking into account
+        various application state parameters.
+        '''
+        return int(self.chars_per_frame * frames)
+
+    def to_frames(self, chars):
+        '''The inverse of to_chars()
+        '''
+        return int(chars / self.chars_per_frame)
 
 
 class DiffApp:
-    def __init__(self, filename_a, filename_b):
-        self.filename_a = filename_a
-        self.filename_b = filename_b
-        self._psf_a = None
-        self._psf_b = None
+    def __init__(self, filename_a, filename_b, terminal=None):
+        self.filenames = AB(filename_a, filename_b)
+        self._psfs = None
+        self._durations = None
         self._diff = None
-        self._len = None
+        self._len = 0
+
+        self._zoom = 1.0
+        self._cues_are_free = True
+        self._start_cue = 0
+        self._end_cue = 0
+        self._active_cue = None
+        self._cursor = 0
+        self._status = ''
+
+        self._draw_state = None
 
         self._loop = asyncio.get_event_loop()
-        self._terminal = LinePrintingTerminal()
-        self._common_fmt = self._terminal.yellow
+        self._terminal = terminal or LinePrintingTerminal()
         self._insertion_fmt = self._terminal.green
+        self._change_fmt = self._terminal.yellow
+        self._cursor_fmt = self._terminal.on_bright_black
         self._loop.add_reader(self._terminal.infile, self._handle_input)
         self._keyboard = Keyboard()
-        self.bindings = {}
+        self.bindings = {
+            '+': self._zoom_in,
+            '=': self._zoom_in,
```

Comment:
What is this binding for? If it is just to prevent frustration in the case where shift wasn't pressed, shouldn't there be an equivalent one for underscore?